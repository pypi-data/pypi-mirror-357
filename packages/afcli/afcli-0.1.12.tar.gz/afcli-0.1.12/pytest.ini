[pytest]
testpaths = tests
python_files = test_*.py
python_classes = Test*
python_functions = test_*
addopts = 
    --verbose
    --cov=afcli
    --cov-report=term-missing
    --cov-report=html:htmlcov
    --cov-fail-under=0
    --strict-markers
    --disable-warnings
markers =
    unit: Unit tests that don't require external dependencies
    integration: Integration tests that may require external services
    slow: Tests that take a long time to run
filterwarnings =
    ignore::DeprecationWarning
    ignore::PendingDeprecationWarning