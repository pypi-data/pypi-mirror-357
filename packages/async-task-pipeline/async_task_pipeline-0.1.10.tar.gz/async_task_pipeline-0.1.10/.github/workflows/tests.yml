# Simple workflow for deploying static content to GitHub Pages
name: Run tests

on:
  # Runs on pushes targeting the default branch
  push:

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

permissions:
  contents: read
  id-token: write

concurrency:
  group: "tests"
  cancel-in-progress: false

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Install uv
        uses: astral-sh/setup-uv@v5
        with:
          enable-cache: true
          cache-dependency-glob: "pyproject.toml"
      - name: Install dependencies
        run: uv sync
      - name: Run tests
        run: make test-cov
