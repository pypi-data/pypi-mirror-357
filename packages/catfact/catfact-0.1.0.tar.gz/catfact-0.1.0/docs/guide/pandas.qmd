---
title: Pandas
jupyter: python3
---


```{python}
import catfact as fct  # for functions
import catfact.pandas  # for .fct methods

from catfact.pandas.data import starwars
```


```{python}
# approach 1: functions
fct.cats(fct.infreq(starwars["eye_color"]))

# approach 2: .fct namespace methods
starwars["eye_color"].fct.infreq().fct.cats()
```

## Ordered flag


```{python}
from catfact.pandas.data import ratings

LEVELS = ["very-bad", "bad", "neutral", "good", "very-good"]


ratings2 = ratings.assign(
    rating_lvl=lambda d: d["rating"].fct.relevel(*LEVELS),
    rating_ord=lambda d: d["rating_lvl"].fct.factor(ordered=True)
)
```

```{python}
# | layout-nrow: 1
from plotnine import ggplot, aes, geom_bar

ggplot(ratings2, aes("rating_lvl", fill="rating_lvl")) + geom_bar()
ggplot(ratings2, aes("rating_ord", fill="rating_ord")) + geom_bar()
```