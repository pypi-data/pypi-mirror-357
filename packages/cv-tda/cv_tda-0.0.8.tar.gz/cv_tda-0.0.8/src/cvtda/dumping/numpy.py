import os
import typing

import numpy

import cvtda.logging

from .base import BaseDumper

class NumpyDumper(BaseDumper[numpy.ndarray]):
    def __init__(self, directory: str):
        self.directory_ = directory

    def get_file_name_(self, name: str):
        return os.path.join(self.directory_, f"{name}.npy")
    
    def execute(self, function: typing.Callable[[typing.Any], numpy.ndarray], name: str, *function_args) -> numpy.ndarray:
        if self.has_dump(name):
            return self.get_dump(name)
        result = function(*function_args)
        self.save_dump(result, name)
        return result

    def save_dump(self, data: numpy.ndarray, name: str):
        file = self.get_file_name_(name)
        cvtda.logging.logger().print(f"Saving the result to {file}")
        os.makedirs(os.path.dirname(file), exist_ok = True)
        numpy.save(file, data)

    def has_dump(self, name: str) -> bool:
        return os.path.exists(self.get_file_name_(name))
    
    def get_dump_impl_(self, name: str) -> numpy.ndarray:
        file = self.get_file_name_(name)
        cvtda.logging.logger().print(f"Got the result from {file}")
        return numpy.load(self.get_file_name_(name))
