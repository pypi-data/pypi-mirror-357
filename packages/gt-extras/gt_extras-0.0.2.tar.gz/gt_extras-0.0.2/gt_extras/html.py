from __future__ import annotations
from typing import Literal

__all__ = ["gt_hyperlink", "with_tooltip"]

# TODO: Add example to docstrings in this module?

def gt_hyperlink(text: str, url: str, new_tab: bool = True) -> int:
    """
    Create HTML hyperlinks for use in `GT` object cells.

    The `gt_hyperlink()` function creates properly formatted HTML hyperlink elements that can be
    used within table cells.

    Parameters
    ----------
    text
        A string that will be displayed as the clickable link text.

    url
        A string indicating the destination URL for the hyperlink.

    new_tab
        A boolean indicating whether the link should open in a new browser tab or the current tab.

    Returns
    -------
    str
        An string containing the HTML formatted hyperlink element.
    """
    target = "_self"
    if new_tab:
        target = "_blank"

    return f'<a href="{url}" target="{target}">{text}</a>'


def with_tooltip(
    label: str,
    tooltip: str,
    text_decoration_style: Literal["solid", "dotted"] | None = "dotted",
    color: str | None = "blue",
) -> str:
    """
    Create HTML text with tooltip functionality for use in GT table cells.

    The `with_tooltip()` function creates an HTML `<abbr>` element with a tooltip that appears
    when users hover over the text. The text can be styled with customizable underline styles
    and colors to indicate it's interactive.

    Parameters
    ----------
    label
        A string that will be displayed as the visible text.

    tooltip
        A string that will appear as the tooltip when hovering over the label.

    text_decoration_style
        A string indicating the style of underline decoration. Options are `"solid"`,
        `"dotted"`, or `None`. If nothing is provided, then `"dotted"` will be used as a default.

    color
        A string indicating the text color. If `None`, no color styling is applied.
        If nothing is provided, then `"blue"` will be used as a default.

    Returns
    -------
    str
        An HTML string containing the formatted tooltip element.
    """

    # Throw if `text_decoration_style` is not one of the three allowed values
    if text_decoration_style not in [None, "solid", "dotted"]:
        raise ValueError("Text_decoration_style must be one of `None`, 'solid', or 'dotted'")

    style = "cursor: help; "

    if text_decoration_style is not None:
        style += "text-decoration: underline; "
        style += f"text-decoration-style: {text_decoration_style}; "
    else:
        style += "text-decoration: none; "

    if color is not None:
        style += f"color: {color}; "

    return f'<abbr style="{style}" title="{tooltip}">{label}</abbr>'
