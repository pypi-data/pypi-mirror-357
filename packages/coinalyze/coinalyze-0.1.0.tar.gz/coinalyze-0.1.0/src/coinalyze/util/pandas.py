import pandas as pd

from coinalyze.constants import COLS_MAPPING
from coinalyze.enums import HistoryEndpoint


def history_response_to_df(
    response: list[dict], endpoint: HistoryEndpoint, index_col: str | None = None
) -> pd.DataFrame:
    """
    Parse a history response to a DataFrame.

    Args:
        response: The JSON response from an history endpoint.
        endpoint: The endpoint to use for the mapping.
        index_col: The column to use as the index.
    """
    dfs = []
    for r in response:
        df = pd.DataFrame(r["history"])
        cols_mapping = {"t": "timestamp"}
        # Depending on the endpoint decide which mapping to use
        cols_mapping.update(COLS_MAPPING[endpoint])
        df = df.rename(columns=cols_mapping)
        df["timestamp"] = pd.to_datetime(df["timestamp"], unit="s")
        df.insert(0, "symbol", r["symbol"])
        if index_col:
            df = df.set_index(index_col)

        dfs.append(df)

    if len(dfs) == 0:
        raise ValueError("No data found")

    df = pd.concat(dfs, axis=0)
    if index_col is None:
        df = df.reset_index(drop=True)

    return df


def response_to_df(response: list[dict], data_type: str | None = None) -> pd.DataFrame:
    """Parse a Coinalyze response to a DataFrame."""
    df = pd.DataFrame(response)
    for col in ["expire_at", "update"]:
        if col in df.columns:
            df[col] = pd.to_datetime(df[col], unit="ms")
    if data_type:
        df.insert(0, "data_type", data_type)
    return df
