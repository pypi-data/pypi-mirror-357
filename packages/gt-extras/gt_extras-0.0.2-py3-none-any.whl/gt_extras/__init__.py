# Import objects from the module
from .themes import (
    gt_theme_538,
    gt_theme_espn,
    gt_theme_nytimes,
    gt_theme_guardian,
    gt_theme_excel,
    gt_theme_dot_matrix,
    gt_theme_dark,
    gt_theme_pff,
)

from .colors import gt_highlight_cols, gt_hulk_col_numeric, gt_color_box

from .icons import fa_icon_repeat, gt_fa_rating

from .plotting import gt_plt_bar, gt_plt_dot

from .html import gt_hyperlink, with_tooltip

__all__ = [
    "gt_theme_538",
    "gt_theme_espn",
    "gt_theme_nytimes",
    "gt_theme_guardian",
    "gt_theme_excel",
    "gt_theme_dot_matrix",
    "gt_theme_dark",
    "gt_theme_pff",

    "gt_highlight_cols",
    "gt_hulk_col_numeric",
    "gt_color_box",

    "fa_icon_repeat",

    "gt_plt_bar",
    "gt_plt_dot",
    
    "gt_hyperlink",
    "with_tooltip",
]
