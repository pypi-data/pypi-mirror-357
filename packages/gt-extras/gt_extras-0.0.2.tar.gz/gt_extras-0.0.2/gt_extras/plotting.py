from __future__ import annotations
from typing import Literal
import warnings

from great_tables import GT
from great_tables._tbl_data import SelectExpr, is_na
from great_tables._locations import resolve_cols_c

from gt_extras._utils_column import (
    _validate_and_get_single_column,
    _scale_numeric_column,
)

from great_tables._data_color.palettes import GradientPalette
from great_tables._data_color.constants import DEFAULT_PALETTE, ALL_PALETTES
from great_tables._data_color.base import (
    _html_color,
    _rescale_factor,
    _get_domain_factor,
)

from svg import SVG, Line, Rect, Text


__all__ = ["gt_plt_bar", "gt_plt_dot"]

# TODO: keep_columns - this is tricky because we can't copy cols in the gt object, so we will have
# to handle the underlying _tbl_data.

# TODO: default font for labels?

# TODO: how to handle negative values? Plots can't really have negative length


def gt_plt_bar(
    gt: GT,
    columns: SelectExpr | None = None,
    fill: str = "purple",
    bar_height: int = 20,
    height: int = 30,
    width: int = 60,
    stroke_color: str | None = "black",
    scale_type: Literal["percent", "number"] | None = None,
    scale_color: str = "white",
    domain: list[int] | list[float] | None = None,
    # keep_columns: bool = False,
) -> GT:
    """
    Create horizontal bar plots in `GT` cells.

    The `gt_plt_bar()` function takes an existing `GT` object and adds horizontal bar charts to
    specified columns. Each cell value is represented as a horizontal bar with length proportional
    to the cell's numeric value relative to the column's maximum value.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    columns
        The columns to target. Can be a single column name or a list of column names. If `None`,
        the bar plot is applied to all numeric columns.

    fill
        The fill color for the bars.

    bar_height
        The height of each individual bar in pixels.

    height
        The height of the bar plot in pixels.

    width
        The width of the maximum bar in pixels

    stroke_color
        The color of the vertical axis on the left side of the bar. The default is black, but if
        `None` is passed no stroke will be drawn.

    scale_type
        The type of value to show on bars. Options are `"number"`, `"percent"`, or `None` for no
        labels.

    scale_color
        The color of text labels on the bars (when `scale_type` is not `None`).

    Returns
    -------
    GT
        A `GT` object with horizontal bar plots added to the specified columns.

    Examples
    --------

    ```{python}
    from great_tables import GT, style, loc
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars.iloc[0:8, list(range(0, 3)) + list(range(5, 11))]

    gt = (
        GT(gtcars_mini,rowname_col="model")
        .tab_stubhead(label="Car")
        .tab_style(style=style.css("text-align: center;"), locations=loc.column_labels())
    )

    gte.gt_plt_bar(gt, columns=["hp", "hp_rpm", "trq", "trq_rpm", "mpg_c", "mpg_h"])
    ```

    Note
    --------
    Each column's bars are scaled independently based on that column's min/max values.
    """
    # A version with svg.py

    if bar_height > height:
        bar_height = height
        warnings.warn(
            f"Bar_height must be less than or equal to the plot height. Adjusting bar_height to {bar_height}.",
            category=UserWarning,
        )

    if bar_height < 0:
        bar_height = 0
        warnings.warn(
            f"Bar_height cannot be negative. Adjusting bar_height to {bar_height}.",
            category=UserWarning,
        )

    # Helper function to make the individual bars
    def _make_bar_html(
        scaled_val: int,
        original_val: int,
        fill: str,
        bar_height: int,
        height: int,
        width: int,
        stroke_color: str,
        scale_type: Literal["percent", "number"] | None,
        scale_color: str,
    ) -> str:
        UNITS = "px" # TODO: let use control this?

        text = ""
        if scale_type == "percent":
            text = str(round(original_val * 100)) + "%"
        if scale_type == "number":
            text = original_val

        canvas = SVG(
            width=str(width) + UNITS,
            height=str(height) + UNITS,
            elements=[
                Rect(
                    x=0,
                    y=str((height - bar_height) / 2) + UNITS,
                    width=str(width * scaled_val) + UNITS,
                    height=str(bar_height) + UNITS,
                    fill=fill,
                    # onmouseover="this.style.fill= 'blue';",
                    # onmouseout=f"this.style.fill='{fill}';",
                ),
                Text(
                    text=text,
                    x=str((width * scaled_val) * 0.98) + UNITS,
                    y=str(height / 2) + UNITS,
                    fill=scale_color,
                    font_size=bar_height * 0.6,
                    text_anchor="end",
                    dominant_baseline="central",
                ),
                Line(
                    x1=0,
                    x2=0,
                    y1=0,
                    y2=str(height) + UNITS,
                    stroke_width=str(height / 10) + UNITS,
                    stroke=stroke_color,
                ),
            ],
        )
        return f'<div style="display: flex;">{canvas.as_str()}</div>'

    # Allow the user to hide the vertical stroke
    if stroke_color is None:
        stroke_color = "#FFFFFF00"

    def make_bar(scaled_val: int, original_val: int) -> str:
        return _make_bar_html(
            scaled_val=scaled_val,
            original_val=original_val,
            fill=fill,
            bar_height=bar_height,
            height=height,
            width=width,
            stroke_color=stroke_color,
            scale_type=scale_type,
            scale_color=scale_color,
        )

    # Get names of columns
    columns_resolved = resolve_cols_c(data=gt, expr=columns)

    res = gt
    for column in columns_resolved:
        # Validate this is a single column and get values
        col_name, col_vals = _validate_and_get_single_column(
            gt,
            column,
        )

        scaled_vals = _scale_numeric_column(gt._tbl_data, col_name, col_vals, domain)

        # Apply the scaled value for each row, so the bar is proportional
        for i, scaled_val in enumerate(scaled_vals):
            res = res.fmt(
                lambda original_val, scaled_val=scaled_val: make_bar(
                    original_val=original_val,
                    scaled_val=scaled_val,
                ),
                columns=column,
                rows=[i],
            )
    return res


def gt_plt_dot(
    gt: GT,
    category_col: SelectExpr,
    data_col: SelectExpr,
    domain: list[int] | list[float] | None = None,
    palette: list[str] | str | None = None,
) -> GT:
    """
    Create dot plots with thin horizontal bars in `GT` cells.

    The `gt_plt_dot()` function takes an existing `GT` object and adds dot plots with horizontal
    bar charts to a specified category column. Each cell displays a colored dot with the category
    label and a horizontal bar representing the corresponding numeric value from the data column.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    category_col
        The column containing category labels that will be displayed next to colored dots.

    data_col
        The column containing numeric values that will determine the length of the horizontal bars.

    domain
        The domain of values to use for the color scheme. This can be a list of floats or integers.
        If `None`, the domain is automatically set to `[0, max(data_col)]`.

    palette
        The color palette to use. This should be a list of colors
        (e.g., `["#FF0000", "#00FF00", "#0000FF"]`). A ColorBrewer palette could also be used,
        just supply the name (see [`GT.data_color()`](https://posit-dev.github.io/great-tables/reference/GT.data_color.html#great_tables.GT.data_color) for additional reference).
        If `None`, then a default palette will be used.

    Returns
    -------
    GT
        A `GT` object with dot plots and horizontal bars added to the specified category column.

    Examples
    --------
    ```{python}
    from great_tables import GT, style, loc
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars.loc[8:20, ["model", "mfr", "hp", "trq", "mpg_c"]]

    gt = (
        GT(gtcars_mini, rowname_col="model")
        .tab_stubhead(label="Car")
    )

    gte.gt_plt_dot(gt, category_col="mfr", data_col="hp")
    ```
    """
    # Get the underlying Dataframe
    data_table = gt._tbl_data

    def _make_bottom_bar_html(
        val: float,
        fill: str,
    ) -> str:
        scaled_value = val * 100
        inner_html = f' <div style="background:{fill}; width:{scaled_value}%; height:4px; border-radius:2px;"></div>'
        html = f'<div style="flex-grow:1; margin-left:0px;"> {inner_html} </div>'

        return html

    def _make_dot_and_bar_html(
        bar_val: float,
        fill: str,
        dot_category_label: str,
    ) -> str:
        if is_na(data_table, bar_val) or is_na(data_table, dot_category_label):
            return "<div></div>"

        label_div_style = "display:inline-block; float:left; margin-right:0px;"

        dot_style = (
            f"height:0.7em; width:0.7em; background-color:{fill};"
            "border-radius:50%; margin-top:4px; display:inline-block;"
            "float:left; margin-right:2px;"
        )

        padding_div_style = (
            "display:inline-block; float:right; line-height:20px; padding:0px 2.5px;"
        )

        bar_container_style = "position:relative; top:1.2em;"

        html = f'''
        <div>
            <div style="{label_div_style}">
                {dot_category_label}
                <div style="{dot_style}"></div>
                <div style="{padding_div_style}"></div>
            </div>
            <div style="{bar_container_style}">
                <div>{_make_bottom_bar_html(bar_val, fill=fill)}</div>
            </div>
        </div>
        '''

        return html.strip()

    # Validate and get data column
    data_col_name, data_col_vals = _validate_and_get_single_column(
        gt,
        data_col,
    )

    # Process numeric data column
    scaled_data_vals = _scale_numeric_column(
        data_table, data_col_name, data_col_vals, domain
    )

    # Validate and get category column
    category_col_name, category_col_vals = _validate_and_get_single_column(
        gt,
        category_col,
    )

    # If palette is not provided, use a default palette
    if palette is None:
        palette = DEFAULT_PALETTE

    # Otherwise get the palette from great_tables._data_color
    elif isinstance(palette, str):
        palette = ALL_PALETTES.get(palette, [palette])

    # Standardize values in `palette` to hexadecimal color values
    palette = _html_color(colors=palette)

    # Rescale the category column for the purpose of assigning colors to each dot
    category_domain = _get_domain_factor(df=data_table, vals=category_col_vals)
    scaled_category_vals = _rescale_factor(
        df=data_table, vals=category_col_vals, domain=category_domain, palette=palette
    )

    # Create a color scale function from the palette
    color_scale_fn = GradientPalette(colors=palette)

    # Call the color scale function on the scaled categoy values to get a list of colors
    color_vals = color_scale_fn(scaled_category_vals)

    # Apply gt.fmt() to each row individually, so we can access the data_value for that row
    res = gt
    for i in range(len(data_table)):
        data_val = scaled_data_vals[i]
        color_val = color_vals[i]

        res = res.fmt(
            lambda x, data=data_val, fill=color_val: _make_dot_and_bar_html(
                dot_category_label=x, fill=fill, bar_val=data
            ),
            columns=category_col,
            rows=[i],
        )

    return res
