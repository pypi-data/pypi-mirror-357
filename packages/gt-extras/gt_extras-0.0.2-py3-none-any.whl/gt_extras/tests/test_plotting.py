import pytest
from gt_extras.tests.conftest import assert_rendered_body

import pandas as pd
import numpy as np
from great_tables import GT
from gt_extras import gt_plt_bar, gt_plt_dot


def test_gt_plt_bar_snap(snapshot, mini_gt):
    res = gt_plt_bar(gt=mini_gt, columns="num")

    assert_rendered_body(snapshot, gt=res)


def test_gt_plt_bar(mini_gt):
    html = gt_plt_bar(gt=mini_gt, columns=["num"]).as_raw_html()
    assert html.count("<svg") == 3


def test_gt_plt_bar_bar_height_too_high(mini_gt):
    with pytest.warns(
        UserWarning,
        match="Bar_height must be less than or equal to the plot height. Adjusting bar_height to 567.",
    ):
        html = gt_plt_bar(
            gt=mini_gt, columns=["num"], bar_height=1234, height=567
        ).as_raw_html()

    assert html.count('height="567px"') == 6
    assert 'height="1234px"' not in html


def test_gt_plt_bar_bar_height_too_low(mini_gt):
    with pytest.warns(
        UserWarning,
        match="Bar_height cannot be negative. Adjusting bar_height to 0.",
    ):
        html = gt_plt_bar(
            gt=mini_gt, columns=["num"], bar_height=-345, height=1234
        ).as_raw_html()

    assert html.count('height="1234px"') == 3
    assert 'height="-345px"' not in html


def test_gt_plt_bar_scale_percent(mini_gt):
    html = gt_plt_bar(gt=mini_gt, columns=["num"], scale_type="percent").as_raw_html()
    assert html.count("%</text>") == 3


def test_gt_plt_bar_scale_number(mini_gt):
    html = gt_plt_bar(gt=mini_gt, columns=["num"], scale_type="number").as_raw_html()
    assert ">33.33</text>" in html


def test_gt_plt_bar_scale_none(mini_gt):
    html = gt_plt_bar(gt=mini_gt, columns=["num"], scale_type=None).as_raw_html()
    assert "</text>" not in html


def test_gt_plt_bar_no_stroke_color(mini_gt):
    html = gt_plt_bar(gt=mini_gt, columns=["num"], stroke_color=None).as_raw_html()
    assert html.count("#FFFFFF00") == 3


def test_gt_plt_bar_type_error(mini_gt):
    with pytest.raises(TypeError, match="Invalid column type provided"):
        gt_plt_bar(gt=mini_gt, columns=["char"]).as_raw_html()


def test_gt_plt_dot_snap(snapshot, mini_gt):
    res = gt_plt_dot(gt=mini_gt, category_col="fctr", data_col="currency")

    assert_rendered_body(snapshot, gt=res)


def test_gt_plt_dot_basic(mini_gt):
    html = gt_plt_dot(gt=mini_gt, category_col="char", data_col="num").as_raw_html()

    # Should contain dot styling
    assert "border-radius:50%; margin-top:4px; display:inline-block;" in html
    assert "height:0.7em; width:0.7em;" in html

    # Should contain bar styling
    assert "flex-grow:1; margin-left:0px;" in html
    assert "width:100.0%; height:4px; border-radius:2px;" in html


# TODO: remove when test_gt_plt_dot_with_palette_xfail() passes.
def test_gt_plt_dot_with_palette(mini_gt):
    html = gt_plt_dot(
        gt=mini_gt,
        category_col="char",
        data_col="num",
        palette=["#FF0000", "#00FF00", "#0000FF"],
    ).as_raw_html()

    assert "#ff0000" in html


@pytest.mark.xfail(reason="Palette bug, issue #717 in great_tables")
def test_gt_plt_dot_with_palette_xfail(mini_gt):
    html = gt_plt_dot(
        gt=mini_gt,
        category_col="char",
        data_col="num",
        palette=["#FF0000", "#00FF00", "#0000FF"],
    ).as_raw_html()

    assert "#ff0000" in html
    assert "#00ff00" in html
    assert "#0000ff" in html


def test_gt_plt_dot_with_domain_expanded(mini_gt):
    html = gt_plt_dot(
        gt=mini_gt, category_col="char", data_col="num", domain=[0, 100]
    ).as_raw_html()

    assert "width:0.1111%; height:4px; border-radius:2px;" in html
    assert "width:2.222%; height:4px; border-radius:2px;" in html
    assert "width:33.33%; height:4px; border-radius:2px;" in html


def test_gt_plt_dot_with_domain_restricted(mini_gt):
    with pytest.warns(
        UserWarning,
        match="Value 33.33 in column 'num' is greater than the domain maximum 10. Setting to 10.",
    ):
        html = gt_plt_dot(
            gt=mini_gt, category_col="char", data_col="num", domain=[0, 10]
        ).as_raw_html()

    assert "width:1.111%; height:4px; border-radius:2px;" in html
    assert "width:22.220000000000002%; height:4px; border-radius:2px;" in html
    assert "width:100%; height:4px; border-radius:2px;" in html


def test_gt_plt_dot_invalid_data_col(mini_gt):
    with pytest.raises(KeyError, match="Column 'invalid_col' not found"):
        gt_plt_dot(gt=mini_gt, category_col="char", data_col="invalid_col")


def test_gt_plt_dot_invalid_category_col(mini_gt):
    with pytest.raises(KeyError, match="Column 'invalid_col' not found"):
        gt_plt_dot(gt=mini_gt, category_col="invalid_col", data_col="num")


def test_gt_plt_dot_multiple_data_cols(mini_gt):
    with pytest.raises(
        ValueError, match="Expected a single column, but got multiple columns"
    ):
        gt_plt_dot(gt=mini_gt, category_col="char", data_col=["num", "char"])


def test_gt_plt_dot_multiple_category_cols(mini_gt):
    with pytest.raises(
        ValueError, match="Expected a single column, but got multiple columns"
    ):
        gt_plt_dot(gt=mini_gt, category_col=["char", "num"], data_col="num")


def test_gt_plt_dot_non_numeric_data_col(mini_gt):
    with pytest.raises(TypeError, match="Invalid column type provided"):
        gt_plt_dot(gt=mini_gt, category_col="char", data_col="char")


def test_gt_plt_dot_with_na_values():
    df = pd.DataFrame(
        {
            "category": ["A", "B", "C", "D"],
            "values": [10, np.nan, 20, None],
        }
    )
    gt = GT(df)

    result = gt_plt_dot(gt=gt, category_col="category", data_col="values")
    html = result.as_raw_html()

    assert isinstance(result, GT)
    assert "width:100.0%; height:4px; border-radius:2px;" in html
    assert html.count("width:0%; height:4px; border-radius:2px;") == 2


def test_gt_plt_dot_with_na_in_category():
    df = pd.DataFrame(
        {
            "category": [np.nan, "B", None, None],
            "values": [5, 10, 10, 5],
        }
    )
    gt = GT(df)

    result = gt_plt_dot(gt=gt, category_col="category", data_col="values")
    html = result.as_raw_html()

    assert isinstance(result, GT)
    assert html.count("width:100.0%; height:4px; border-radius:2px;") == 1
    assert "width:50.0%; height:4px; border-radius:2px;" not in html


def test_gt_plt_dot_palette_string_valid(mini_gt):
    html = gt_plt_dot(
        gt=mini_gt, category_col="char", data_col="num", palette="viridis"
    ).as_raw_html()

    assert "background:#440154;" in html
