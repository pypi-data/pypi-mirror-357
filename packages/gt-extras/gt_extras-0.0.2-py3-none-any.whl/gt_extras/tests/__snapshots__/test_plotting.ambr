# serializer version: 1
# name: test_gt_plt_bar_snap
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right"><div style="display: flex;"><svg xmlns="http://www.w3.org/2000/svg" width="60px" height="30px"><rect x="0" y="5.0px" width="0.2px" height="20px" fill="purple"/><text dominant-baseline="central" text-anchor="end" font-size="12.0" fill="white" x="0.196px" y="15.0px"/><line stroke="black" stroke-width="3.0px" x1="0" y1="0" x2="0" y2="30px"/></svg></div></td>
      <td class="gt_row gt_left">apricot</td>
      <td class="gt_row gt_left">one</td>
      <td class="gt_row gt_right">2015-01-15</td>
      <td class="gt_row gt_right">13:35</td>
      <td class="gt_row gt_right">2018-01-01 02:22</td>
      <td class="gt_row gt_right">49.95</td>
      <td class="gt_row gt_left">row_1</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right"><div style="display: flex;"><svg xmlns="http://www.w3.org/2000/svg" width="60px" height="30px"><rect x="0" y="5.0px" width="4.0px" height="20px" fill="purple"/><text dominant-baseline="central" text-anchor="end" font-size="12.0" fill="white" x="3.92px" y="15.0px"/><line stroke="black" stroke-width="3.0px" x1="0" y1="0" x2="0" y2="30px"/></svg></div></td>
      <td class="gt_row gt_left">banana</td>
      <td class="gt_row gt_left">two</td>
      <td class="gt_row gt_right">2015-02-15</td>
      <td class="gt_row gt_right">14:40</td>
      <td class="gt_row gt_right">2018-02-02 14:33</td>
      <td class="gt_row gt_right">17.95</td>
      <td class="gt_row gt_left">row_2</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right"><div style="display: flex;"><svg xmlns="http://www.w3.org/2000/svg" width="60px" height="30px"><rect x="0" y="5.0px" width="60.0px" height="20px" fill="purple"/><text dominant-baseline="central" text-anchor="end" font-size="12.0" fill="white" x="58.8px" y="15.0px"/><line stroke="black" stroke-width="3.0px" x1="0" y1="0" x2="0" y2="30px"/></svg></div></td>
      <td class="gt_row gt_left">coconut</td>
      <td class="gt_row gt_left">three</td>
      <td class="gt_row gt_right">2015-03-15</td>
      <td class="gt_row gt_right">15:45</td>
      <td class="gt_row gt_right">2018-03-03 03:44</td>
      <td class="gt_row gt_right">1.39</td>
      <td class="gt_row gt_left">row_3</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
  </tbody>
  '''
# ---
# name: test_gt_plt_dot_snap
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
      <td class="gt_row gt_left"><div>
              <div style="display:inline-block; float:left; margin-right:0px;">
                  one
                  <div style="height:0.7em; width:0.7em; background-color:#000000;border-radius:50%; margin-top:4px; display:inline-block;float:left; margin-right:2px;"></div>
                  <div style="display:inline-block; float:right; line-height:20px; padding:0px 2.5px;"></div>
              </div>
              <div style="position:relative; top:1.2em;">
                  <div><div style="flex-grow:1; margin-left:0px;">  <div style="background:#000000; width:100.0%; height:4px; border-radius:2px;"></div> </div></div>
              </div>
          </div></td>
      <td class="gt_row gt_right">2015-01-15</td>
      <td class="gt_row gt_right">13:35</td>
      <td class="gt_row gt_right">2018-01-01 02:22</td>
      <td class="gt_row gt_right">49.95</td>
      <td class="gt_row gt_left">row_1</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
      <td class="gt_row gt_left"><div>
              <div style="display:inline-block; float:left; margin-right:0px;">
                  two
                  <div style="height:0.7em; width:0.7em; background-color:#4cbd81;border-radius:50%; margin-top:4px; display:inline-block;float:left; margin-right:2px;"></div>
                  <div style="display:inline-block; float:right; line-height:20px; padding:0px 2.5px;"></div>
              </div>
              <div style="position:relative; top:1.2em;">
                  <div><div style="flex-grow:1; margin-left:0px;">  <div style="background:#4cbd81; width:35.93593593593593%; height:4px; border-radius:2px;"></div> </div></div>
              </div>
          </div></td>
      <td class="gt_row gt_right">2015-02-15</td>
      <td class="gt_row gt_right">14:40</td>
      <td class="gt_row gt_right">2018-02-02 14:33</td>
      <td class="gt_row gt_right">17.95</td>
      <td class="gt_row gt_left">row_2</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
      <td class="gt_row gt_left"><div>
              <div style="display:inline-block; float:left; margin-right:0px;">
                  three
                  <div style="height:0.7em; width:0.7em; background-color:#9653ca;border-radius:50%; margin-top:4px; display:inline-block;float:left; margin-right:2px;"></div>
                  <div style="display:inline-block; float:right; line-height:20px; padding:0px 2.5px;"></div>
              </div>
              <div style="position:relative; top:1.2em;">
                  <div><div style="flex-grow:1; margin-left:0px;">  <div style="background:#9653ca; width:2.7827827827827822%; height:4px; border-radius:2px;"></div> </div></div>
              </div>
          </div></td>
      <td class="gt_row gt_right">2015-03-15</td>
      <td class="gt_row gt_right">15:45</td>
      <td class="gt_row gt_right">2018-03-03 03:44</td>
      <td class="gt_row gt_right">1.39</td>
      <td class="gt_row gt_left">row_3</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
  </tbody>
  '''
# ---
